File: 10181-puzzle/10181-puzzle.cpp
#include <bits/stdc++.h>
using namespace std;

const int dr[4] = {0, 0, +1, -1};
const int dc[4] = {+1, -1, 0, 0};
const int dir[4] = {'R', 'L', 'D', 'U'};
const int INF = 0x3f3f3f3f;
const int FOUND = -1;
vector<char> path;
int A[15][15], Er, Ec;

int H() {
    int h = 0;
    for (int r = 0; r < 4; r++) {
        for (int c = 0; c < 4; c++) {
            if (A[r][c] == 0) continue;
            int expect_r = (A[r][c] - 1) / 4;
            int expect_c = (A[r][c] - 1) % 4;
            //printf("A[i][j]%d  expect_r %d expect_C %d\n",A[r][c],expect_r,expect_c);
            h += abs(expect_r - r) + abs(expect_c - c);
        }
    }
    //printf("en H h %d\n",h);
    return h;
}

int dfs(int g, int pdir, int bound) {
    int h = H();
    //printf("%d\n",h);
    int f = g + h;
    if (f > bound) return f;
    if (h == 0) return FOUND;

    int mn = INF;
    for (int i = 0; i < 4; i++) {
        if (i == (pdir ^ 1)) continue;

        int nr = Er + dr[i];
        int nc = Ec + dc[i];
        if (nr < 0 || nr >= 4) continue;
        if (nc < 0 || nc >= 4) continue;

        path.push_back(dir[i]);
        swap(A[nr][nc], A[Er][Ec]);
        swap(nr, Er); swap(nc, Ec);
        int t = dfs(g + 1, i, bound);
        if (t == FOUND) return FOUND;
        if (t < mn) mn = t;
        swap(nr, Er); swap(nc, Ec);
        swap(A[nr][nc], A[Er][Ec]);
        path.pop_back();
    }

    return mn;
}

bool IDAstar() {
    int bound = H();
    for (;;) {
    	//printf("Bound %d\n", bound);
        int t = dfs(0, -1, bound);
        //printf("%d %d %d\n",Er,Ec,t);
        if (t == FOUND) return true;
        if (t == INF) return false;
        // ????? bound >= 50,?????? >= 50,??
        if (t >= 50) return false;
        bound = t;
    }
    return false;
}

bool solvable() {
    int cnt = 0;
    bool occur[16] = {false};
    for (int r = 0; r < 4; r++) {
        for (int c = 0; c < 4; c++) {
            if (A[r][c] == 0) {
                Er = r;
                Ec = c;
            }
            else {
            	//printf(" cnt %d Er %d Ec %d  x %d\n",cnt,Er,Ec,A[r][c]);
                cnt += count(occur + 1, occur + A[r][c], false);
                occur[A[r][c]] = true;
            }
        }
    }
    //printf("Er %d Ec %d\n",Er,Ec);
    return (cnt + (Er + 1)) % 2 == 0;
}

int main() {
    int TC;
    scanf("%d", &TC);
    while (TC--) {
        for (int r = 0; r < 4; r++) {
            for (int c = 0; c < 4; c++) {
                scanf("%d", &A[r][c]);
            }
        }

        path.clear();

        if (!solvable() || !IDAstar()) {
            puts("This puzzle is not solvable.");
        }
        else {
            for (int i = 0; i < path.size(); i++ ) {
                printf("%c", path[i]);
            }
            puts("");
        }
    }

    return 0;
}

File: 10181-puzzle.py
from sys import stdin

dr = [0,0,1,-1]
dc = [1,-1,0,0]
dire = ["R", "L", "D", "U"]
INF = 2**31 - 1
found = -1
A = [[] for i in range(4) ]
Er = 0
Ec = 0
path = []

def H():
    h = 0
    for i in range(4):
        for j in range(4):
            if A[i][j] == 0:
                continue
            expect_r = ((A[i][j]-1) // 4)
            expect_C = ((A[i][j]-1) % 4)
            h += abs(expect_r - i) + abs(expect_C - j)
            #print("A[i][j]",A[i][j],"expect_r",expect_r,"expect_C",expect_C)
    #print("en H h",h)
    return h

def dfs(g, pdir, bound):
    global Er,Ec
    h = H()
    #print("h",h)
    f = g + h
    if( f > bound ): return f
    if( h == 0): return found
    mn = INF
    for i in range(4):
        if( i == (pdir ^ 1)):
            continue
        nr = Er + dr[i]
        nc = Ec + dc[i]
        if( nr < 0 or nr >= 4): continue
        if( nc < 0 or nc >= 4): continue
        path.append(dire[i])
        cop = A[nr][nc]
        A[nr][nc] = A[Er][Ec]
        A[Er][Ec] = cop
        cop = nr
        nr = Er
        Er = cop
        cop = nc
        nc = Ec
        Ec = cop
        t = dfs(g+1, i, bound)
        if( t == found ): return found
        if( t < mn ): mn = t
        cop = nr
        nr = Er
        Er = cop
        cop = nc
        nc = Ec
        Ec = cop
        cop = A[nr][nc]
        A[nr][nc] = A[Er][Ec]
        A[Er][Ec] = cop
        path.pop()
    return mn


def IDAstar():
    global Er,Ec,path
    bound = H()
    while( True ):
        #print("Bound",bound)
        t = dfs(0,-1, bound)
        #print("Er Ec t",Er,Ec,t)
        if( t == found ): return True
        if( t == INF ): return False
        if( t>= 50 ): return False
        bound = t

def solvable():
    global Er,Ec,path
    cnt = 0
    occur = [False] * 16
    for i in range(4):
        for j in range(4):
            if( A[i][j] == 0 ):
                Er = i
                Ec = j
            else:
                x = A[i][j]
                #print(" cnt",cnt,"Er",Er,"Ec",Ec,"X",x )
                cnt += occur[1:x].count(False)
                occur[x] = True
    #print("Er",Er,"Ec",Ec)
    return (cnt + (Er + 1))%2 == 0

def main():
    global Er,Ec,path
    cases = int(stdin.readline().strip())
    for ca in range(cases):
        for i in range(4):
            A[i] = [int(x) for x in stdin.readline().strip().split()]
        
        path = []
        #for i in A:
        #   print(i)
        #"""
        #x  = solvable()
        #y = IDAstar()
        #print(x,y,path)
        if( not solvable() or not IDAstar()):
            print("This puzzle is not solvable.")
        else:
            print("".join(path))
        #"""
main()

















